src/Sparrow/Utils/ZstdStream.cs

```diff
@@ -181,60 +201,104 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
+            WaitOnSemaphoreOrThrowIfDisposed();
 
-            while (true)
+            try
+            {
+                while (true)
+                {
+                    var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
+                    if (outputBytes == 0)
+                        break;
+                    _inner.Write(_tempBuffer, 0, outputBytes);
+                }
+                _inner?.Flush();
+            }
+            finally
             {
-                var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
-                if (outputBytes == 0)
-                    break;
-                _inner.Write(_tempBuffer, 0, outputBytes);
+                _semaphoreSlim.Release();
             }
-            _inner?.Flush();
         }
 
         public override async Task FlushAsync(CancellationToken cancellationToken)
         {
-            if (_disposeOnce.Disposed)
-                throw new ObjectDisposedException("Object was already disposed!");
+            await WaitOnSemaphoreOrThrowIfDisposedAsync().ConfigureAwait(false);
 
-            while (true)
+            try
             {
-                var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
-                if (outputBytes == 0)
-                    break;
+                while (true)
+                {
+                    var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
+                    if (outputBytes == 0)
+                        break;
 
-                await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+                    await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+                }
+                await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
             }
-            await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
+            finally
+            {
+                _semaphoreSlim.Release();
+            }
+        }
+
+        private void WaitOnSemaphoreOrThrowIfDisposed()
+        {
+            if (_disposeOnce.Disposed)
+                throw new ObjectDisposedException("Object was already disposed!");
+
+            _semaphoreSlim.Wait();
+        }
+
+        private async Task WaitOnSemaphoreOrThrowIfDisposedAsync()
+        {
+            if (_disposeOnce.Disposed)
+                throw new ObjectDisposedException("Object was already disposed!");
+
+            await _semaphoreSlim.WaitAsync().ConfigureAwait(false);
         }
 
         private void DisposeInternal()
         {
+            if (_semaphoreSlim.CurrentCount <= 0)
+            {
+                // ensure we are done reading/writing
+                _semaphoreSlim.Release();
```

Comment:
How does it ensure that? It looks like if you will hit Dispose in the middle of read/write you can release that and continue with Disposal. Is this what we want? It suggest that we did not release a semaphore somewhere.

Also I suggest doing:
```
ValueTask<ReleaseSemaphore> WaitOnSemaphoreOrThrowIfDisposedAsync();

struct ReleaseSemaphore : IDisposable
{
}
```

and use usings for code readability.